### DISCUSSION TITLE
Issue with creating custom exception with custom fields

### DISCUSSION THREAD

**Discussion by faxioman** [PyO3/pyo3#3259] [Questions] [answered]
I'm facing an issue while trying to create a custom exception in PyO3 with custom data.

I have looked at the implementation of `pyextend` as explained in this https://github.com/PyO3/pyo3/issues/2597.

When I try to import my module I get the error:

```
pyo3_runtime.PanicException: Failed to initialize new exception type.: PyErr { type: <class 'TypeError'>, value: TypeError("type 'builtins.MyException' is not an acceptable base type"), traceback: None }
```

This is my exception implementation:

```
use pyo3::create_exception;
use pyo3::prelude::*;


#[pyclass(extends=pyo3::exceptions::PyException)]
pub struct MyException {
    #[pyo3(get)]
    pub context: &'static str,
}

#[pymethods]
impl MyException {
    #[new]
    fn new() -> Self {
        MyException {
            context: "Hello :)",
        }
    }
}

create_exception!(sls_messaging.exceptions, SomethingWentWrongException, MyException);
```

And I'm adding the exception in this way:

```
let exceptions_mod = PyModule::new(_py, "exceptions")?;
exceptions_mod.add("SomethingWentWrongException", _py.get_type::<SomethingWentWrongException>())?;
m.add_submodule(exceptions_mod)?;
```

Ideas? Thanks!!

**Comment by faxioman** [comment, marked as answer]
Ok, I think I've found a way but without using the create_exception! macro.
This is the custom exception definition:

```
use pyo3::exceptions::PyException;
use pyo3::prelude::*;

#[pyclass(extends=PyException)]
pub struct SomethingWentWrongException {
    #[pyo3(get, set)]
    pub data: String,
}

#[pymethods]
impl SomethingWentWrongException {
    #[new]
    fn new(data: &str) -> Self {
        SomethingWentWrongException { data: data.to_string() }
    }
}
```

And this how I'm creating a new instance of it:

```
let exception = PyErr::new::<SomethingWentWrongException, _>(("Custom data string here").to_object(py));
Err(exception)
```

**Comment by birkenfeld** [comment]
In general if that works could we make it the preferred way to implement new exceptions? (@davidhewitt @adamreichold) Maybe a bit of special sauce in `pyclass` to recognize `PyException` bases?

**Reply by davidhewitt** [reply]
I'm very keen to rip out `create_exception` - the discussion is at #295 and simply it's not been my priority to sit down and ensure that the design pieces of  `#[pyclass]` are sufficient to justify the removal of `create_exception` quite yet.
